### ISSUE TITLE
[API Proposal]: Allow specifying indent size and whitespace character when writing JSON with Utf8JsonWriter

### ISSUE THREAD

**Issue by noqn** [dotnet/runtime#63882]
### Background and motivation

This proposal would enable Utf8JsonWriter to write indented JSON with either spaces or tabs, and with a variable number of whitespace characters per indentation.

To preface, indentation size was brought up around two years ago in #1174. Although the area owners were open for input, there were little discussion at the time and the issue was closed. Now that more eyes are on System.Text.Json I am hoping to revisit the topic.

___

As is well known, people form strong opinions about indentation. In a context where the serialized JSON will be inspected or manually edited by the users, there will inevitably be a demand for the application to conform to their indentation preferences. Adding this functionality to System.Text.Json would in turn let application developers provide indentation settings to their users.

This would also be useful for when the serialized JSON must match the indentation of the source files which were deserialized e.g., for version control purposes.

### API Proposal

```C#
namespace System.Text.Json
{
    public sealed class JsonSerializerOptions 
    {
        public char IndentCharacter { get; set; }
        public int IndentSize { get; set; }
    }
    
    public struct JsonWriterOptions 
    {
        public char IndentCharacter { get; set; }
        public int IndentSize { get; set; }
    }
}
```

`IndentCharacter` would define which whitespace character is used and `IndentSize` would define the number of whitespace characters that would be written per indentation level.

To match the behavior of `JsonWriterOptions.MaxDepth`, the default values of `IndentCharacter` (`'\0'`) and `IndentSize` (`0`) would represent fallback values (which would be properly assigned in the `Utf8JsonWriter` constructor).

### API Usage

``` C#
var jsonWriterOptions = new JsonWriterOptions() 
{
    Indented = true,
    IndentCharacter = '\t',
    IndentSize = 1
};

var jsonSerializerOptions = new JsonSerializerOptions()
{
    WriteIndented = true,
    IndentSize = 4
};
```

### Alternative Designs

For the sake of comparison, this is how Json.NET does it:

``` C#
namespace Newtonsoft.Json 
{
    public class JsonTextWriter ...
    {
        public int Indentation { get; set; }
        public char IndentChar { get; set; }
    }
}
```

Json.NET's ```Indentation``` property combines the behavior of ```Indented``` and ```IndentSize``` by using the default value of 0 to disable formating altogether. Such a solution would however clash with our current API.

**Comment by eiriktsarpalis** [comment]
> Json.NET defines the indentation whitespace character with a char property. Since spaces and tabs are the only valid (non-newline) JSON whitespace characters, we can more elegantly cover all allowed values with a simple boolean.

I personally find the Json.NET API design to be cleaner, but we'd need to check if we can use that without sacrificing performance. We also need to make sure that the existing `Indented` property interacts correctly with the proposed `IndentSize` property (setting one should generally update the other).

> In https://github.com/dotnet/runtime/issues/1174, the area owners of System.Text.Json understandably voiced concern over introducing complexity or performance degradation to the library. I would like to demonstrate that this proposal is achievable without such compromises.

Would it be possible to share benchmark results comparing performance of a prototype vs the current implementation?

Slightly related: #54715

**Comment by noqn** [comment]
> > Json.NET defines the indentation whitespace character with a char property. Since spaces and tabs are the only valid (non-newline) JSON whitespace characters, we can more elegantly cover all allowed values with a simple boolean.
> 
> I personally find the Json.NET API design to be cleaner, but we'd need to check if we can use that without sacrificing performance. 

If we move away from `bool IndentWithTabs`, should we use a `byte` or `char` property for the indent character?

- `byte IndentCharacter` would be preferable for performance since `Utf8JsonWriter` could use it as-is without casting.
- `char IndentCharacter` on the other hand is clearer and more convenient for the user.

```csharp
// char IndentCharacter
var options = new JsonWriterOptions
{
    Indented = true,
    IndentCharacter = '\t'
};
```

```csharp
// byte IndentCharacter
var options = new JsonWriterOptions
{
    Indented = true,
    IndentCharacter = (byte) '\t',
};
```

A third option is a `char` property in the public API that is backed by an `internal` `byte` property.
For performance, `Utf8JsonWriter` could access the byte property as-is without casting, and the public API would be cleaner.

It could be implemented something like this:

```csharp
public struct JsonWriterOptions
{
    internal byte IndentByte { get; private set; }

    public char IndentCharacter
    { 
        get => (char)IndentByte;
        set
        {
            if (value is not ' ' and not '\t' and not '\0') 
            {
                throw new ArgumentException("...");
            }

            IndentByte = (byte)value;
        }
    }
}
```

Perhaps a terrible idea, idk.

>We also need to make sure that the existing `Indented` property interacts correctly with the proposed `IndentSize` property (setting one should generally update the other).

I think this might introduce confusion, for example

```csharp
var options = new JsonWriterOptions();

options.Indented = true;  // IndentSize: 2 (JsonConstants.SpacesPerIndent)
options.IndentSize = 4;   // IndentSize: 4
options.Indented = false; // IndentSize: 0?
options.Indented = true;  // IndentSize: 2?
```

It would be preferable if setting one does not affect the other. If `Indented` is false, the value of `IndentSize` won't matter anyways. 

> Would it be possible to share benchmark results comparing performance of a prototype vs the current implementation?
> 
> Slightly related: #54715

I ran the Utf8JsonWriter suite in dotnet/performance: https://github.com/dotnet/performance/tree/main/src/benchmarks/micro/libraries/System.Text.Json/Utf8JsonWriter

Here's the output of the ResultComparison tool

```
68 executed benchmarks 

summary:
better: 8, geomean: 1.036
worse: 10, geomean: 1.177
total diff: 18
```

| Slower                                                                           | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Text.Json.Tests.Perf_Ctor.Ctor(Formatted: False, SkipValidation: False)   |      1.46 |            15.00 |            21.83 |         |
| System.Text.Json.Tests.Perf_Ctor.Ctor(Formatted: True, SkipValidation: False)    |      1.44 |            15.25 |            22.03 |         |
| System.Text.Json.Tests.Perf_Ctor.Ctor(Formatted: False, SkipValidation: True)    |      1.42 |            15.27 |            21.70 |         |
| System.Text.Json.Tests.Perf_Ctor.Ctor(Formatted: True, SkipValidation: True)     |      1.41 |            15.44 |            21.83 |         |
| System.Text.Json.Tests.Perf_Booleans.WriteBooleans(Formatted: False, SkipValidat |      1.04 |       1596606.25 |       1667521.88 |         |
| System.Text.Json.Tests.Perf_Strings.WriteStringsUtf16(Formatted: True, SkipValid |      1.04 |       7734531.25 |       8055377.42 |         |
| System.Text.Json.Tests.Perf_Strings.WriteStringsUtf16(Formatted: False, SkipVali |      1.03 |      12569154.17 |      12942183.33 |         |
| System.Text.Json.Tests.Perf_Strings.WriteStringsUtf16(Formatted: False, SkipVali |      1.03 |       6868138.89 |       7044231.43 |         |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf16(Formatted: True, SkipValidatio |      1.02 |       2100920.17 |       2151054.78 |         |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf8(Formatted: False, SkipValidatio |      1.02 |       1606743.59 |       1644223.53 |         |

| Faster                                                                           | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Text.Json.Tests.Perf_Booleans.WriteBooleans(Formatted: True, SkipValidati |      1.05 |       2115260.16 |       2007234.77 |         |
| System.Text.Json.Tests.Perf_Strings.WriteStringsUtf8(Formatted: True, SkipValida |      1.05 |       4736417.31 |       4510072.73 |         |
| System.Text.Json.Tests.Perf_Strings.WriteStringsUtf8(Formatted: True, SkipValida |      1.04 |       9439903.85 |       9036607.69 |         |
| System.Text.Json.Tests.Perf_Basic.WriteBasicUtf16(Formatted: False, SkipValidati |      1.04 |       1652555.92 |       1592353.50 |         |
| System.Text.Json.Tests.Perf_Strings.WriteStringsUtf8(Formatted: True, SkipValida |      1.03 |      63244625.00 |      61240750.00 |         |
| System.Text.Json.Tests.Perf_Strings.WriteStringsUtf8(Formatted: True, SkipValida |      1.03 |       4789055.77 |       4666719.81 |         |
| System.Text.Json.Tests.Perf_Deep.WriteDeepUtf16(Formatted: False, SkipValidation |      1.02 |       4623970.37 |       4514632.73 |         |
| System.Text.Json.Tests.Perf_Strings.WriteStringsUtf8(Formatted: False, SkipValid |      1.02 |       8717124.14 |       8523317.24 |         |


The `Utf8JsonWriter` constructor is slower due to setting its `JsonWriterOptions`' `IndentSize` and `IndentCharacter` to fallback values if they were left at `default` (what is done to `JsonWriterOptions.MaxDepth`).

The remaining 14 benchmarks with a performance difference >2% (out of 68 total) are split fairly evenly between slower and faster than the main branch. I presume my system is not producing very stable results in those benchmarks...

I imagined that the WriteDeepUtf8 benchmark might reflect accurately on changes to indentation logic, so I ran it again on the main branch, two branches which implements `IndentSize` and `IndentCharacter` respectively, and one which implements both.


|        Method     | Branch                      | Formatted    | SkipValidation     | Mean Ratio |          Mean |         Error |        StdDev | Median Ratio |        Median |           Min |           Max |     Allocated |
|------------------ |-----------------------------|------------- |------------------- |------------|--------------:|--------------:|--------------:|--------------|--------------:|--------------:|--------------:|--------------:|
| **WriteDeepUtf8** | **Main**                    |      False   |            False   | **1.000**  |   4.724 ms    |   0.0111 ms   |   0.0104 ms   | **1.000**    |    4.721 ms   |    4.708 ms   |    4.740 ms   |       345 B   |
| **WriteDeepUtf8** | **IndentChar + IndentSize** |      False   |            False   | **0.981**  |   4.632 ms    |   0.0089 ms   |   0.0083 ms   | **0,981**    |    4.630 ms   |    4.621 ms   |    4.646 ms   |       353 B   |
| **WriteDeepUtf8** | **IndentSize**              |      False   |            False   | **0.968**  |   4.571 ms    |   0.0057 ms   |   0.0053 ms   | **0,968**    |    4.570 ms   |    4.562 ms   |    4.581 ms   |       353 B   |
| **WriteDeepUtf8** | **IndentChar**              |      False   |            False   | **0.963**  |   4.547 ms    |   0.0090 ms   |   0.0080 ms   | **0,963**    |    4.546 ms   |    4.538 ms   |    4.567 ms   |       353 B   |

|            Method | Branch                      | Formatted    | SkipValidation     | Mean Ratio |          Mean |         Error |        StdDev | Median Ratio |        Median |           Min |           Max |     Allocated |
|------------------ |-----------------------------|------------- |------------------- |------------|--------------:|--------------:|--------------:|--------------|--------------:|--------------:|--------------:|--------------:|
| **WriteDeepUtf8** | **Main**                    |      False   |             True   | **1.000**  |   4.269 ms    |   0.0045 ms   |   0.0042 ms   | **1.000**    |    4.269 ms   |    4.260 ms   |    4.277 ms   |       128 B   |
| **WriteDeepUtf8** | **IndentChar + IndentSize** |      False   |             True   | **1.005**  |   4.292 ms    |   0.0057 ms   |   0.0054 ms   | **1,006**    |    4.293 ms   |    4.284 ms   |    4.301 ms   |       136 B   |
| **WriteDeepUtf8** | **IndentSize**              |      False   |             True   | **1.036**  |   4.423 ms    |   0.0090 ms   |   0.0084 ms   | **1,036**    |    4.424 ms   |    4.410 ms   |    4.438 ms   |       148 B   |
| **WriteDeepUtf8** | **IndentChar**              |      False   |             True   | **1.051**  |   4.485 ms    |   0.0064 ms   |   0.0057 ms   | **1,050**    |    4.484 ms   |    4.475 ms   |    4.493 ms   |       137 B   |

|        Method     | Branch                      | Formatted    | SkipValidation     | Mean Ratio |          Mean |         Error |        StdDev | Median Ratio |        Median |           Min |           Max |     Allocated |
|--------------     |--------                     |----------    |---------------     |------------|--------------:|    ----------:|    ----------:|--------------|    ----------:|    ----------:|    ----------:|    ----------:|
| **WriteDeepUtf8** | **Main**                    |       True   |            False   | **1.000**  |   41.355 ms   |   5.9859 ms   |   6.8933 ms   | **1.000**    |   36.945 ms   |   36.344 ms   |   56.320 ms   |       480 B   |
| **WriteDeepUtf8** | **IndentChar + IndentSize** |       True   |            False   | **0.992**  |   41.013 ms   |   6.3328 ms   |   7.2929 ms   | **0,985**    |   36.394 ms   |   35.840 ms   |   57.018 ms   |       824 B   |
| **WriteDeepUtf8** | **IndentSize**              |       True   |            False   | **1.002**  |   41.450 ms   |   5.8866 ms   |   6.7790 ms   | **1,004**    |   37.086 ms   |   36.137 ms   |   51.664 ms   |       488 B   |
| **WriteDeepUtf8** | **IndentChar**              |       True   |            False   | **0.973**  |   40.231 ms   |   4.4017 ms   |   5.0690 ms   | **0,990**    |   36.563 ms   |   35.930 ms   |   47.736 ms   |       824 B   |

|            Method | Branch                      | Formatted    | SkipValidation     | Mean Ratio |          Mean |         Error |        StdDev | Median Ratio |        Median |           Min |           Max |     Allocated |
|------------------ |-----------------------------|------------- |---------------     |------------|--------------:|--------------:|--------------:|--------------|--------------:|--------------:|--------------:|--------------:|
| **WriteDeepUtf8** | **Main**                    |       True   |             True   | **1.000**  |   37.772 ms   |   1.8500 ms   |   2.1305 ms   | **1.000**    |   36.752 ms   |   35.870 ms   |   43.240 ms   |       600 B   |
| **WriteDeepUtf8** | **IndentChar + IndentSize** |       True   |             True   | **1.081**  |   40.835 ms   |   5.4765 ms   |   6.3067 ms   | **0,988**    |   36.303 ms   |   36.083 ms   |   50.221 ms   |       608 B   |
| **WriteDeepUtf8** | **IndentSize**              |       True   |             True   | **1.084**  |   40.940 ms   |   5.3766 ms   |   6.1917 ms   | **1,007**    |   36.997 ms   |   36.669 ms   |   50.426 ms   |       608 B   |
| **WriteDeepUtf8** | **IndentChar**              |       True   |             True   | **1.090**  |   41.183 ms   |   5.9331 ms   |   6.8325 ms   | **1,004**    |   36.904 ms   |   36.165 ms   |   55.615 ms   |       608 B   |

Thoughts?

**Comment by ds5678** [comment]
+1

**Comment by noqn** [comment]
@layomia @eiriktsarpalis any more input?

Feel free to view the implementation here: https://github.com/noqn/runtime/commit/e7a73b3cc469edd83d4e8205db6283ff55b3c20f

Current proposal:

```diff
namespace System.Text.Json
{
     public sealed class JsonSerializerOptions 
     {
+         public char IndentCharacter { get; set; }
+         public int IndentSize { get; set; }
     }

     public struct JsonWriterOptions 
     {
+         public char IndentCharacter { get; set; }
+         public int IndentSize { get; set; }
     }
}
```

Is there any chance this could be added to #63762? It seems realistic to implement in time for .NET 7

I ran some more benchmarks on a modified and lighter version of the WriteDeepUtf8 benchmarks, hoping to get more stable results. Let me know if you want me to post the full benchmarks. Otherwise, are we on the same page that there are no remaining performance concerns?

|     Depth |     Formatted |     SkipValidation |            Mean |         Median |      Mean (Old) |   Median (Old) | Mean ratio | Median ratio |
|---------- |-------------- |------------------- |----------------:|---------------:|----------------:|---------------:|------------|--------------|
|    **10** |     **False** |          **False** |    **338,2 ns** |   **337,9 ns** |    **344,4 ns** |   **344,2 ns** | 0,98      | 0,98 
|    **10** |     **False** |           **True** |    **309,1 ns** |   **308,9 ns** |    **295,5 ns** |   **295,2 ns** | 1,05      | 1,05 
|    **10** |      **True** |          **False** |    **532,2 ns** |   **532,1 ns** |    **542,6 ns** |    **542,7 ns** | 0,98       | 0,98 
|    **10** |      **True** |           **True** |    **480,9 ns** |   **480,2 ns** |    **504,5 ns** |    **504,5 ns** | 0,95       | 0,95 
|   **100** |     **False** |          **False** |  **2963,8 ns** |  **2963,1 ns** |   **2969,1 ns** |   **2969,3 ns**| 1,00       | 1,00 
|   **100** |     **False** |           **True** |  **2296,5 ns** |  **2297,1 ns** |  **2232,7 ns** |  **2232,3 ns** | 1,03       | 1,03 
|   **100** |      **True** |           **True** |  **4782,9 ns** |  **4782,8 ns** |  **4997,2 ns** |  **4997,1 ns** | 0,96       | 0,96 
|   **500** |     **False** |          **False** | **15168,8 ns** | **15170,3 ns** | **15121,4 ns** | **15121,9 ns** | 1,00       | 1,00
|   **500** |     **False** |           **True** | **11031,5 ns** | **11033,1 ns** | **11042,4 ns** | **11037,2 ns** | 1,00       | 1,00
|   **500** |      **True** |          **False** | **54533,1 ns** | **54558,6 ns** | **55916,3 ns** | **55911,6 ns** | 0,98       | 0,98
|   **500** |      **True** |           **True** | **51802,2 ns** | **51782,4 ns** | **53183,4 ns** | **53175,4 ns** | 0,97       | 0,97

**Comment by noqn** [comment]
> > Let me know if you want me to post the full benchmarks.
> 
> Yes please. Please also share benchmark results [using statistical tests](https://github.com/dotnet/performance/blob/main/src/benchmarks/micro/README.md#private-runtime-builds). We need to make sure no regressions are introduced in `Utf8JsonWriter` since it's part of the serialization hot path.

Thanks, here're the WriteDeep benchmarks results with [DataSize](https://github.com/dotnet/performance/blob/main/src/benchmarks/micro/libraries/System.Text.Json/Utf8JsonWriter/Perf.Deep.cs#L16) decreased (from 100_000) to 10_000 for more stable results:

____

|         Method |        Job |                                                                                                                 Toolchain | Formatted | SkipValidation |       Mean |    Error |   StdDev |     Median |        Min |        Max | Ratio | MannWhitney(2%) | Allocated | Alloc Ratio |
|--------------- |----------- |-------------------------------------------------------------------------------------------------------------------------- |---------- |--------------- |-----------:|---------:|---------:|-----------:|-----------:|-----------:|------:|---------------- |----------:|------------:|
|  **WriteDeepUtf8** | **Job-ZOSJRV** | **\dotnet-runtime-upstream\\...\corerun.exe** |     **False** |          **False** |   **473.2 μs** |  **1.43 μs** |  **1.26 μs** |   **472.9 μs** |   **471.6 μs** |   **475.9 μs** |  **1.00** |            **Base** |     **337 B** |        **1.00** |
|  WriteDeepUtf8 | Job-KXWMUA |          \dotnet-runtime\\...\corerun.exe |     False |          False |   475.2 μs |  0.42 μs |  0.39 μs |   475.2 μs |   474.6 μs |   475.8 μs |  1.00 |            Same |     346 B |        1.03 |
|                |            |                                                                                                                           |           |                |            |          |          |            |            |            |       |                 |           |             |
| WriteDeepUtf16 | Job-ZOSJRV | \dotnet-runtime-upstream\\...\corerun.exe |     False |          False |   508.3 μs |  0.58 μs |  0.51 μs |   508.4 μs |   507.5 μs |   509.3 μs |  1.00 |            Base |     337 B |        1.00 |
| WriteDeepUtf16 | Job-KXWMUA |          \dotnet-runtime\\...\corerun.exe |     False |          False |   485.1 μs |  0.94 μs |  0.88 μs |   484.8 μs |   483.8 μs |   486.8 μs |  0.95 |          Faster |     345 B |        1.02 |
|                |            |                                                                                                                           |           |                |            |          |          |            |            |            |       |                 |           |             |
|  **WriteDeepUtf8** | **Job-ZOSJRV** | **\dotnet-runtime-upstream\\...\corerun.exe** |     **False** |           **True** |   **434.5 μs** |  **0.54 μs** |  **0.48 μs** |   **434.6 μs** |   **433.8 μs** |   **435.2 μs** |  **1.00** |            **Base** |     **120 B** |        **1.00** |
|  WriteDeepUtf8 | Job-KXWMUA |          \dotnet-runtime\\...\corerun.exe |     False |           True |   440.0 μs |  0.61 μs |  0.51 μs |   440.0 μs |   439.2 μs |   440.8 μs |  1.01 |            Same |     129 B |        1.07 |
|                |            |                                                                                                                           |           |                |            |          |          |            |            |            |       |                 |           |             |
| WriteDeepUtf16 | Job-ZOSJRV | \dotnet-runtime-upstream\\...\corerun.exe |     False |           True |   452.9 μs |  0.50 μs |  0.44 μs |   453.0 μs |   451.9 μs |   453.5 μs |  1.00 |            Base |     121 B |        1.00 |
| WriteDeepUtf16 | Job-KXWMUA |          \dotnet-runtime\\...\corerun.exe |     False |           True |   454.7 μs |  0.55 μs |  0.52 μs |   454.8 μs |   453.7 μs |   455.5 μs |  1.00 |            Same |     129 B |        1.07 |
|                |            |                                                                                                                           |           |                |            |          |          |            |            |            |       |                 |           |             |
|  **WriteDeepUtf8** | **Job-ZOSJRV** | **\dotnet-runtime-upstream\\...\corerun.exe** |      **True** |          **False** | **2,777.1 μs** | **14.37 μs** | **13.44 μs** | **2,775.2 μs** | **2,754.8 μs** | **2,799.0 μs** |  **1.00** |            **Base** |     **341 B** |        **1.00** |
|  WriteDeepUtf8 | Job-KXWMUA |          \dotnet-runtime\\...\corerun.exe |      True |          False | 2,771.5 μs | 13.58 μs | 12.70 μs | 2,768.8 μs | 2,753.5 μs | 2,794.0 μs |  1.00 |            Same |     356 B |        1.04 |
|                |            |                                                                                                                           |           |                |            |          |          |            |            |            |       |                 |           |             |
| WriteDeepUtf16 | Job-ZOSJRV | \dotnet-runtime-upstream\\...\corerun.exe |      True |          False | 2,770.8 μs |  9.08 μs |  8.49 μs | 2,772.1 μs | 2,758.6 μs | 2,782.5 μs |  1.00 |            Base |     348 B |        1.00 |
| WriteDeepUtf16 | Job-KXWMUA |          \dotnet-runtime\\...\corerun.exe |      True |          False | 2,773.4 μs | 10.95 μs | 10.24 μs | 2,769.9 μs | 2,762.4 μs | 2,792.4 μs |  1.00 |            Same |     356 B |        1.02 |
|                |            |                                                                                                                           |           |                |            |          |          |            |            |            |       |                 |           |             |
|  **WriteDeepUtf8** | **Job-ZOSJRV** | **\dotnet-runtime-upstream\\...\corerun.exe** |      **True** |           **True** | **2,761.1 μs** | **14.45 μs** | **13.52 μs** | **2,763.8 μs** | **2,743.0 μs** | **2,782.1 μs** |  **1.00** |            **Base** |     **122 B** |        **1.00** |
|  WriteDeepUtf8 | Job-KXWMUA |          \dotnet-runtime\\...\corerun.exe |      True |           True | 2,763.5 μs | 11.53 μs | 10.79 μs | 2,767.3 μs | 2,743.9 μs | 2,779.9 μs |  1.00 |            Same |     133 B |        1.09 |
|                |            |                                                                                                                           |           |                |            |          |          |            |            |            |       |                 |           |             |
| WriteDeepUtf16 | Job-ZOSJRV | \dotnet-runtime-upstream\\...\corerun.exe |      True |           True | 2,764.1 μs | 10.26 μs |  9.60 μs | 2,764.5 μs | 2,748.2 μs | 2,782.8 μs |  1.00 |            Base |     132 B |        1.00 |
| WriteDeepUtf16 | Job-KXWMUA |          \dotnet-runtime\\...\corerun.exe |      True |           True | 2,761.1 μs |  9.51 μs |  8.90 μs | 2,759.8 μs | 2,749.1 μs | 2,774.9 μs |  1.00 |            Same |     133 B |        1.01 |

**Comment by eiriktsarpalis** [comment]
Thanks. We can take a look at this in API review. Could you update your OP with updated API proposal?

**Comment by noqn** [comment]
> Thanks. We can take a look at this in API review. Could you update your OP with updated API proposal?

Awesome, I've updated the proposal.

**Comment by noqn** [comment]
@eiriktsarpalis Can this be tagged with `api-ready-for-review` or did I miss something?

**Comment by eiriktsarpalis** [comment]
@noqn done.

**Comment by amercer-fdi** [comment]
Any movement on this? It's a strange inconsistency that when I open Visual Studio it formats JSON to 4 spaces but I can only Serialize to 2 spaces.

**Comment by noqn** [comment]
@eiriktsarpalis can this be moved to the 8.0 milestone? It has been stuck in the API review backlog for 9 months due to being in "future". Thanks.

**Comment by bartonjs** [comment]
* Because `char`+`int` is limiting (it doesn't allow anything with varied indentation, like "->  ", or anything outside the Basic Multilingual Plane) we combined them into one field.
* While we typed the field as `string`, `JsonEncodedText` might be more appropriate
* The property is proposed as non-nullable, defaulting to whatever the current indentation is (two spaces? four?)

```C#
namespace System.Text.Json
{
    public sealed class JsonSerializerOptions 
    {
        public string IndentText { get; set; }
    }
    
    public struct JsonWriterOptions 
    {
        public string IndentText { get; set; }
    }
}
```

**Comment by noqn** [comment]
Hi, thanks for reviewing 😄 


>     * Because `char`+`int` is limiting (it doesn't allow anything with varied indentation, like "->  ", or anything outside the Basic Multilingual Plane) we combined them into one field.

I have major concerns about implementing `IndentText` without guardrails. `JsonSerializer` is strict about not writing invalid JSON and I firmly believe the user should only be allowed to indent with valid JSON whitespace characters (tabs and spaces - presuming we exclude CR & LF).

With that in mind, `char IndentCharacter` being limited to BPM is a non-issue if we would've put proper guardrails on `IndentText` anyways. Since that felt like the main argument in favor of the `IndentText` proposal, I think it's worth reconsidering the old proposal.

As someone mentioned in the review, it would be hard to parse (and adjust) the exact indentation with `IndentText`:

```cs
var jsonWriterOptions = new JsonWriterOptions() 
{
    Indented = true,
    IndentText = "        "
};
```

As opposed to:

```cs
var jsonWriterOptions = new JsonWriterOptions() 
{
    Indented = true,
    IndentSize = 8
};
```

Overall I feel like the old proposal is a more pleasing.

In terms of performance, defining a single indent char allows us to keep utilizing `Span.Fill` when writing the indentations. But this might be negligible, idk.

____

Someone mentioned in the review about defining the indent character as an enum. I like this since compared to the old proposal it makes it more clear to the users which chars are permitted. I'll write this down as an alternative proposal:

```cs
namespace System.Text.Json
{
    public sealed class JsonSerializerOptions 
    {
        public JsonIndentCharacter IndentCharacter { get; set; }
        public int IndentSize { get; set; }
    }
    
    public struct JsonWriterOptions 
    {
        public JsonIndentCharacter IndentCharacter { get; set; }
        public int IndentSize { get; set; }
    }

    public enum JsonIndentCharacter
    {
        Space,
        Tab
    }
}
```

Usage:
```cs
var jsonSerializerOptions = new JsonSerializerOptions()
{
    WriteIndented = true,
    IndentSize = 1
    IndentCharacter = JsonIndentCharacter.Tab
};
```

**Comment by eiriktsarpalis** [comment]
> ```cs
> var jsonWriterOptions = new JsonWriterOptions() 
> {
>     Indented = true,
>     IndentText = "        "
> };
> ```

That certainly seems a bit awkward to me. @bartonjs I think we should reconsider the design here.

**Comment by bartonjs** [comment]
After a long discussion, we seem to have returned to using a single string property.

We did update the shape to account for the new source generator attribute, though.

```C#
namespace System.Text.Json
{
    public sealed class JsonSerializerOptions 
    {
        public string IndentText { get; set; }
    }
    
    public struct JsonWriterOptions 
    {
        public string IndentText { get; set; }
    }
}

namespace System.Text.Json.Serialization
{
    public partial class JsonSourceGenerationOptionsAttribute
    {
        public string IndentText { get; set; }
    }
}
```

**Comment by Neme12** [comment]
I don't think 0 indent size (or empty `IndentText`) should be disallowed. What if you want to write everything on a separate line, as if it was indented, but without any indentation? That would translate to `WriteIndented` being true and an indent size of 0.

**Comment by terrajobst** [comment]
- Looks good as proposed
- The default is going to be a single space and indent size of 2

```C#
namespace System.Text.Json
{
    public sealed class JsonSerializerOptions
    {
        public char IndentCharacter { get; set; }
        public int IndentSize { get; set; }
    }

    public struct JsonWriterOptions
    {
        public char IndentCharacter { get; set; }
        public int IndentSize { get; set; }
    }
}

namespace System.Text.Json.Serialization
{
    public partial class JsonSourceGenerationOptionsAttribute
    {
        public char IndentCharacter { get; set; }
        public int IndentSize { get; set; }
    }
}
```

**Comment by noqn** [comment]
Not meant to snatch this from under @manandre's nose, but since I had an implementation for the now-approved API laying around for ~2 years I figured I could just brush it off and make a pull request : )

**Comment by eiriktsarpalis** [comment]
Thanks, but I think the nice thing to do here would be to give @manandre a chance to finish their work.

**Comment by noqn** [comment]
Hi. hope I don't come across as too blunt, but you asked me two years ago to implement and benchmark this proposal before it was even allowed for review, _against the API review process guideline_.

I've now published a pull request, which currently sits with all checks passed. Respectfully, there is no reason to wait around for another implementation that would have to be rewritten from scratch anyways. I'd greatly appreciate it if you started reviewing the pull request I've delivered. Thanks.
